File: assignment_4/provided_resource/angular_spectrum_part_1.m
%% Problem 4C solution - angular spectrum propagator
% Author: John Bass
function outputField = paddedAngularSpectrum(inputField,x_max_m,z_m,lambda_m,nPad_pix)
    % Pad input field
    paddedInputField = zeros(size(inputField)+2*nPad_pix);
    paddedInputField(nPad_pix+1:nPad_pix+size(inputField,1),...
                     nPad_pix+1:nPad_pix+size(inputField,2)) = inputField;

    % Save pixel side length of input field, for use later
    N_unpadded = size(inputField,1);
    N = size(paddedInputField,1);

    % Get spatial domain coordinate system
    dx_m = 2*x_max_m / (N_unpadded-1);
    x_m = -(N-1)*dx_m/2:dx_m:(N-1)*dx_m/2;

    % Get fourier-domain coordinate array. Even-length frames go up to a
    % different max frequency, and thus must be treated differently. See:
    % https://www.mathworks.com/help/matlab/ref/fft.html, "Noisy Signal"
    % example
    fx_max_pm = 1/dx_m/2;
    dfx_pm = 1/dx_m/N;
    if(mod(length(x_m),2)==0)
        fx_pm = linspace(-fx_max_pm,fx_max_pm-dfx_pm,N);
    else
        fx_pm = linspace(-fx_max_pm,fx_max_pm,N);
    end

    % Get frequency-domain coordinate matrices
    kx_radpm = 2 * pi * fx_pm;
    [Kx_radpm,Ky_radpm] = meshgrid(kx_radpm,kx_radpm);

    % Get wavenumber
    k_radpm = 2 * pi / lambda_m;

    % Define kernel function
    kernel = exp(1j * z_m * sqrt(k_radpm.^2 - Kx_radpm.^2 - Ky_radpm.^2));

    % Propagate the input field
    paddedInputField_fft = fftshift(fft2(paddedInputField));
    paddedOutputField_fft = paddedInputField_fft .* kernel;
    paddedOutputField = ifft2(ifftshift(paddedOutputField_fft));

    % Unpad input field
    outputField = paddedOutputField(nPad_pix+1:nPad_pix+size(inputField,1),...
                                   nPad_pix+1:nPad_pix+size(inputField,2));
end

% === Propagate point source to observation window (same size as diffuser) ===

% Define physical parameters
lambda_m = 500e-9;         % Wavelength [m]
z_m = 0.5;                 % Propagation distance [m]
x_max_m = 37.5e-3;         % Half-width of the observation window [m] → full window = 75 mm
N = 1001;                  % Resolution of field (un-padded)
nPad_pix = N;              % Padding to prevent wraparound artifacts

% Create spatial grid for field definition
dx_m = 2 * x_max_m / (N - 1);
x_m = linspace(-x_max_m, x_max_m, N);
[X, Y] = meshgrid(x_m, x_m);

% Define field arriving at z = 0 from a point source at z = -z_m
R = sqrt(X.^2 + Y.^2 + z_m^2);
inputField = exp(1j * 2 * pi * R / lambda_m) ./ R;

% Propagate field to distance z_m using original function
outputField = paddedAngularSpectrum(inputField, x_max_m, z_m, lambda_m, nPad_pix);

% Display phase of the propagated field
figure;
imagesc(x_m*1e3, x_m*1e3, angle(outputField));
axis image;
xlabel('x (mm)');
ylabel('y (mm)');
title('Phase at Observation Window (75 mm wide)');
colorbar;

File: assignment_4/provided_resource/angular_spectrum_part_1.py
import numpy as np
from numpy.fft import fft2, ifft2, fftshift, ifftshift

def padded_angular_spectrum(input_field, x_max_m, z_m, lambda_m, n_pad_pix):
    """
    Propagates an optical field using the Angular Spectrum method with padding.
    
    Parameters:
        input_field (ndarray): The input field (amplitude distribution).
        x_max_m (float): Maximum spatial extent (meters).
        z_m (float): Propagation distance (meters).
        lambda_m (float): Wavelength (meters).
        n_pad_pix (int): Number of pixels to pad.

    Returns:
        ndarray: Output field after propagation.
    """
    # Pad input field with zeros
    padded_input_field = np.pad(input_field, n_pad_pix, mode='constant')

    # Get sizes of padded and unpadded fields
    N_unpadded = input_field.shape[0]  # Assuming square input field
    N = padded_input_field.shape[0]  # Padded size

    # Compute spatial coordinates
    dx_m = 2 * x_max_m / (N_unpadded - 1)
    x_m = np.linspace(-(N - 1) * dx_m / 2, (N - 1) * dx_m / 2, N)

    # Compute frequency domain coordinates
    fx_max_pm = 1 / (2 * dx_m)  # Maximum frequency
    dfx_pm = 1 / (N * dx_m)  # Frequency step size

    if len(x_m) % 2 == 0:
        fx_pm = np.linspace(-fx_max_pm, fx_max_pm - dfx_pm, N)  # Even case
    else:
        fx_pm = np.linspace(-fx_max_pm, fx_max_pm, N)  # Odd case

    # Create frequency domain coordinate matrices
    kx_radpm = 2 * np.pi * fx_pm
    Kx_radpm, Ky_radpm = np.meshgrid(kx_radpm, kx_radpm)

    # Compute wavenumber
    k_radpm = 2 * np.pi / lambda_m

    # Compute transfer function (propagation kernel)
    kernel = np.exp(1j * z_m * np.sqrt(k_radpm**2 - Kx_radpm**2 - Ky_radpm**2))

    # Apply Angular Spectrum Method
    padded_input_fft = fftshift(fft2(padded_input_field))  # Forward FFT
    padded_output_fft = padded_input_fft * kernel  # Multiply by transfer function
    padded_output_field = ifft2(ifftshift(padded_output_fft))  # Inverse FFT

    # Crop out the original field size
    output_field = padded_output_field[n_pad_pix:n_pad_pix + N_unpadded,
                                       n_pad_pix:n_pad_pix + N_unpadded]

    return output_field